parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.kt

```diff
@@ -0,0 +1,442 @@
+/*
+    Copyright 2016 Parse.ly, Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+package com.parsely.parselyandroid
+
+import android.content.Context
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.ProcessLifecycleOwner
+import com.parsely.parselyandroid.Logging.log
+import java.util.UUID
+
+/**
+ * Tracks Parse.ly app views in Android apps
+ *
+ *
+ * Accessed as a singleton. Maintains a queue of pageview events in memory and periodically
+ * flushes the queue to the Parse.ly pixel proxy server.
+ */
+public open class ParselyTracker protected constructor(siteId: String, flushInterval: Int, c: Context) {
```

Comment:
This shouldn't be `public` and `open` but I haven't found any other way for now. If I introduced some interface, let's say `Tracker` to use it in `EngagementManagerTest` to create a fake implementation, I'd have to expose `enqueueEvent` to `public` - apparently, interface methods cannot be `internal`.

But I believe this shouldn't be a blocker - `ParselyTracker`, as previously written in Java, was "`open`" by default.

Reply:
>  If I introduced some interface, let's say `Tracker` to use it in `EngagementManagerTest` to create a fake implementation, I'd have to expose `enqueueEvent` to `public` - apparently, interface methods cannot be `internal`.

Assuming that this is what we wanted to do, wouldn't it be more appropriate to use multiple interfaces in this case? We'd have a `Tracker` interface that we expose to `public` and another `internal` interface that has `enqueueEvent` function. `ParselyTracker` then will implement both and everything should work? Let me know if I am missing anything in this scenario.

Reply:
Even with multiple interfaces, we can't have a method in interface that is `internal`. Even if the interface itself is `internal`

<img width="489" alt="image" src="https://github.com/Parsely/parsely-android/assets/5845095/3fe4bd33-1b76-4368-bcb6-7c55f4df1de4">

If we didn't add `internal`, then `ParselyTracker#enqueueEvent` would be exposed to API consumers, which is something we should omit.

Reply:
It's very likely that I am missing something, but I was thinking something like this would work? 74a9f1857b3a0609e39df92e1cb08ee31e1028ba

Note that this ^ is not production quality code, I just put it together to make it easier to discuss. Let me know what you think!

---

Also, if we did go with such a design, I think we can make `ParselyTracker` internal as well and expose the shared instance some other way.

Reply:
Thank you! I really like your idea, haven't seen in this way myself. I also appreciate it clarifies the public contract better.

I've cherry-picked your comment and added some changes. Most notably, I've renamed `Tracker` interface to `ParselyTracker` to minimize breaking changes for the API consumers.

Reply:
I am glad you found it useful! 🥳